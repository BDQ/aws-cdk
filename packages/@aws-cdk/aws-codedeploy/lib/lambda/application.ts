import cdk = require("@aws-cdk/cdk");
import { ApplicationImportProps, ApplicationProps, IApplication } from "../base/application";
import { CfnApplication } from "../codedeploy.generated";
import { applicationNameToArn } from "../utils";

/**
 * A CodeDeploy Application that deploys to an AWS Lambda function.
 */
export class LambdaApplication extends cdk.Construct implements IApplication {
  /**
   * Import an Application defined either outside the CDK,
   * or in a different CDK Stack and exported using the {@link IApplication#export} method.
   *
   * @param scope the parent Construct for this new Construct
   * @param id the logical ID of this new Construct
   * @param props the properties of the referenced Application
   * @returns a Construct representing a reference to an existing Application
   */
  public static import(scope: cdk.Construct, id: string, props: ApplicationImportProps): IApplication {
    return new ImportedLambdaApplication(scope, id, props);
  }

  public readonly applicationArn: string;
  public readonly applicationName: string;

  constructor(scope: cdk.Construct, id: string, props: ApplicationProps = {}) {
    super(scope, id);

    const resource = new CfnApplication(this, 'Resource', {
      applicationName: props.applicationName,
      computePlatform: 'Lambda'
    });

    this.applicationName = resource.ref;
    this.applicationArn = applicationNameToArn(this.applicationName, this);
  }

  public export(): ApplicationImportProps {
    return {
      applicationName: new cdk.Output(this, 'ApplicationName', { value: this.applicationName }).makeImportValue().toString()
    };
  }
}

class ImportedLambdaApplication extends cdk.Construct implements IApplication {
  public readonly applicationArn: string;
  public readonly applicationName: string;

  constructor(scope: cdk.Construct, id: string, private readonly props: ApplicationImportProps) {
    super(scope, id);

    this.applicationName = props.applicationName;
    this.applicationArn = applicationNameToArn(props.applicationName, this);
  }

  public export(): ApplicationImportProps {
    return this.props;
  }
}
