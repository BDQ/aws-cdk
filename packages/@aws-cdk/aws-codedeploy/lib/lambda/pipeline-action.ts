import codepipeline = require('@aws-cdk/aws-codepipeline-api');
import cdk = require('@aws-cdk/cdk');
import { BasePipelineDeployAction, CommonPipelineDeployActionProps } from '../base/pipeline-action';
import { LambdaDeploymentGroup } from './deployment-group';

export interface LambdaPipelineDeployActionProps extends CommonPipelineDeployActionProps {
  deploymentGroup: LambdaDeploymentGroup;
}

export class LambdaPipelineDeployAction extends BasePipelineDeployAction {
  protected readonly deploymentGroup: LambdaDeploymentGroup;

  constructor(props: LambdaPipelineDeployActionProps) {
    super({
      ...props,
      configuration: {
        ApplicationName: props.deploymentGroup.application.applicationName,
        DeploymentGroupName: props.deploymentGroup.deploymentGroupName,
      }
    });

    this.deploymentGroup = props.deploymentGroup;
  }

  protected bind(stage: codepipeline.IStage, scope: cdk.Construct): void {
    this.addCommonPermissions(stage, scope, this.deploymentGroup);
  }
}
