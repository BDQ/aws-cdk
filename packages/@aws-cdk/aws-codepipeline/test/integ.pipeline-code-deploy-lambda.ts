
import codecommit = require('@aws-cdk/aws-codecommit');
import codedeploy = require('@aws-cdk/aws-codedeploy');
import lambda = require('@aws-cdk/aws-lambda');
import cdk = require('@aws-cdk/cdk');
import codepipeline = require('../lib');

import path = require('path');

const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-codepipeline-codedeploy-lambda');
const pipeline = new codepipeline.Pipeline(stack, 'Pipeline');

const repo = new codecommit.Repository(stack, 'Repo', {
  repositoryName: 'Repo',
});

const sourceAction = repo.toCodePipelineSourceAction({ actionName: 'CodeCommit' });
pipeline.addStage({
  name: 'Source',
  actions: [sourceAction],
});

const handler = new lambda.Function(stack, `Handler`, {
  code: lambda.Code.asset(path.join(__dirname, 'lambda/handler')),
  handler: 'index.handler',
  runtime: lambda.Runtime.NodeJS810,
});

const version = handler.addVersion('1');
const blueGreenAlias = new lambda.Alias(stack, `Alias`, {
  aliasName: `alias`,
  version
});

const deploymentGroup = new codedeploy.LambdaDeploymentGroup(stack, 'BlueGreenDeployment', {
  alias: blueGreenAlias,
  deploymentConfig: codedeploy.LambdaDeploymentConfig.Linear10PercentEvery1Minute,
});

// const deployAction = new codedeploy.LambdaPipelineDeployAction({
//   actionName: 'CodeDeploy',
//   inputArtifact: sourceAction.outputArtifact,
//   deploymentGroup,
// });

const deployStage = pipeline.addStage({ name: 'Deploy' });
deployStage.addAction(deploymentGroup.toCodePipelineDeployAction({
  actionName: 'CodeDeploy',
  inputArtifact: sourceAction.outputArtifact,
}));

app.run();
