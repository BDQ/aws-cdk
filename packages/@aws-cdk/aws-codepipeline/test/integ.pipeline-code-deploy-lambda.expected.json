{
  "Resources": {
    "PipelineArtifactsBucket22248F97": {
      "Type": "AWS::S3::Bucket",
      "DeletionPolicy": "Retain"
    },
    "PipelineRoleD68726F7": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codepipeline.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "PipelineRoleDefaultPolicyC7A05455": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*"
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "PipelineArtifactsBucket22248F97",
                    "Arn"
                  ]
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "PipelineArtifactsBucket22248F97",
                          "Arn"
                        ]
                      },
                      "/*"
                    ]
                  ]
                }
              ]
            },
            {
              "Action": [
                "codecommit:GetBranch",
                "codecommit:GetCommit",
                "codecommit:UploadArchive",
                "codecommit:GetUploadArchiveStatus",
                "codecommit:CancelUploadArchive"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "Repo02AC86CF",
                  "Arn"
                ]
              }
            },
            {
              "Action": [
                "codedeploy:GetApplicationRevision",
                "codedeploy:RegisterApplicationRevision"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":codedeploy:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":application:",
                    {
                      "Ref": "BlueGreenDeploymentApplication36C892C0"
                    }
                  ]
                ]
              }
            },
            {
              "Action": [
                "codedeploy:CreateDeployment",
                "codedeploy:GetDeployment"
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":codedeploy:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":deploymentgroup:",
                    {
                      "Ref": "BlueGreenDeploymentApplication36C892C0"
                    },
                    "/",
                    {
                      "Ref": "BlueGreenDeployment5C188134"
                    }
                  ]
                ]
              }
            },
            {
              "Action": "codedeploy:GetDeploymentConfig",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":codedeploy:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":deploymentconfig:CodeDeployDefault.LambdaLinear10PercentEvery1Minute"
                  ]
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "PipelineRoleDefaultPolicyC7A05455",
        "Roles": [
          {
            "Ref": "PipelineRoleD68726F7"
          }
        ]
      }
    },
    "PipelineC660917D": {
      "Type": "AWS::CodePipeline::Pipeline",
      "Properties": {
        "RoleArn": {
          "Fn::GetAtt": [
            "PipelineRoleD68726F7",
            "Arn"
          ]
        },
        "Stages": [
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Source",
                  "Owner": "AWS",
                  "Provider": "CodeCommit",
                  "Version": "1"
                },
                "Configuration": {
                  "RepositoryName": {
                    "Fn::GetAtt": [
                      "Repo02AC86CF",
                      "Name"
                    ]
                  },
                  "BranchName": "master",
                  "PollForSourceChanges": false
                },
                "InputArtifacts": [],
                "Name": "CodeCommit",
                "OutputArtifacts": [
                  {
                    "Name": "Artifact_CodeCommit_awscdkcodepipelinecodedeploylambdaRepoCF6F437B"
                  }
                ],
                "RunOrder": 1
              }
            ],
            "Name": "Source"
          },
          {
            "Actions": [
              {
                "ActionTypeId": {
                  "Category": "Deploy",
                  "Owner": "AWS",
                  "Provider": "CodeDeploy",
                  "Version": "1"
                },
                "Configuration": {
                  "ApplicationName": {
                    "Ref": "BlueGreenDeploymentApplication36C892C0"
                  },
                  "DeploymentGroupName": {
                    "Ref": "BlueGreenDeployment5C188134"
                  }
                },
                "InputArtifacts": [
                  {
                    "Name": "Artifact_CodeCommit_awscdkcodepipelinecodedeploylambdaRepoCF6F437B"
                  }
                ],
                "Name": "CodeDeploy",
                "OutputArtifacts": [],
                "RunOrder": 1
              }
            ],
            "Name": "Deploy"
          }
        ],
        "ArtifactStore": {
          "Location": {
            "Ref": "PipelineArtifactsBucket22248F97"
          },
          "Type": "S3"
        }
      },
      "DependsOn": [
        "PipelineRoleDefaultPolicyC7A05455",
        "PipelineRoleD68726F7"
      ]
    },
    "PipelineEventsRole46BEEA7C": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "PipelineEventsRoleDefaultPolicyFF4FCCE0": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codepipeline:StartPipelineExecution",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition"
                    },
                    ":codepipeline:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":",
                    {
                      "Ref": "PipelineC660917D"
                    }
                  ]
                ]
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "PipelineEventsRoleDefaultPolicyFF4FCCE0",
        "Roles": [
          {
            "Ref": "PipelineEventsRole46BEEA7C"
          }
        ]
      }
    },
    "Repo02AC86CF": {
      "Type": "AWS::CodeCommit::Repository",
      "Properties": {
        "RepositoryName": "Repo",
        "Triggers": []
      }
    },
    "RepoawscdkcodepipelinecodedeploylambdaPipeline40CCE661EventRule3E19C56C": {
      "Type": "AWS::Events::Rule",
      "Properties": {
        "EventPattern": {
          "source": [
            "aws.codecommit"
          ],
          "resources": [
            {
              "Fn::GetAtt": [
                "Repo02AC86CF",
                "Arn"
              ]
            }
          ],
          "detail-type": [
            "CodeCommit Repository State Change"
          ],
          "detail": {
            "event": [
              "referenceCreated",
              "referenceUpdated"
            ],
            "referenceName": [
              "master"
            ]
          }
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition"
                  },
                  ":codepipeline:",
                  {
                    "Ref": "AWS::Region"
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId"
                  },
                  ":",
                  {
                    "Ref": "PipelineC660917D"
                  }
                ]
              ]
            },
            "Id": "Pipeline",
            "RoleArn": {
              "Fn::GetAtt": [
                "PipelineEventsRole46BEEA7C",
                "Arn"
              ]
            }
          }
        ]
      }
    },
    "HandlerServiceRoleFCDC14AE": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "Handler886CB40B": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "HandlerCodeS3Bucket8DD11ED9"
          },
          "S3Key": {
            "Fn::Join": [
              "",
              [
                {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "HandlerCodeS3VersionKey0BB5191E"
                        }
                      ]
                    }
                  ]
                },
                {
                  "Fn::Select": [
                    1,
                    {
                      "Fn::Split": [
                        "||",
                        {
                          "Ref": "HandlerCodeS3VersionKey0BB5191E"
                        }
                      ]
                    }
                  ]
                }
              ]
            ]
          }
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "HandlerServiceRoleFCDC14AE",
            "Arn"
          ]
        },
        "Runtime": "nodejs8.10"
      },
      "DependsOn": [
        "HandlerServiceRoleFCDC14AE"
      ]
    },
    "HandlerVersion1F237F6D0": {
      "Type": "AWS::Lambda::Version",
      "Properties": {
        "FunctionName": {
          "Ref": "Handler886CB40B"
        }
      }
    },
    "Alias325C5727": {
      "Type": "AWS::Lambda::Alias",
      "Properties": {
        "FunctionName": {
          "Ref": "Handler886CB40B"
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "HandlerVersion1F237F6D0",
            "Version"
          ]
        },
        "Name": "alias"
      },
      "UpdatePolicy": {
        "CodeDeployLambdaAliasUpdate": {
          "ApplicationName": {
            "Ref": "BlueGreenDeploymentApplication36C892C0"
          },
          "DeploymentGroupName": {
            "Ref": "BlueGreenDeployment5C188134"
          }
        }
      }
    },
    "BlueGreenDeploymentApplication36C892C0": {
      "Type": "AWS::CodeDeploy::Application",
      "Properties": {
        "ComputePlatform": "Lambda"
      }
    },
    "BlueGreenDeploymentServiceRole225851FB": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codedeploy.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          "arn:aws:iam::aws:policy/service-role/AWSCodeDeployRoleForLambda"
        ]
      }
    },
    "BlueGreenDeployment5C188134": {
      "Type": "AWS::CodeDeploy::DeploymentGroup",
      "Properties": {
        "ApplicationName": {
          "Ref": "BlueGreenDeploymentApplication36C892C0"
        },
        "ServiceRoleArn": {
          "Fn::GetAtt": [
            "BlueGreenDeploymentServiceRole225851FB",
            "Arn"
          ]
        },
        "AutoRollbackConfiguration": {
          "Enabled": true,
          "Events": [
            "DEPLOYMENT_FAILURE"
          ]
        },
        "DeploymentConfigName": "CodeDeployDefault.LambdaLinear10PercentEvery1Minute",
        "DeploymentStyle": {
          "DeploymentOption": "WITH_TRAFFIC_CONTROL",
          "DeploymentType": "BLUE_GREEN"
        }
      }
    }
  },
  "Parameters": {
    "HandlerCodeS3Bucket8DD11ED9": {
      "Type": "String",
      "Description": "S3 bucket for asset \"aws-cdk-codepipeline-codedeploy-lambda/Handler/Code\""
    },
    "HandlerCodeS3VersionKey0BB5191E": {
      "Type": "String",
      "Description": "S3 key for asset version \"aws-cdk-codepipeline-codedeploy-lambda/Handler/Code\""
    }
  }
}